src/espresso-migration/EspressoMigration.sol

```diff
@@ -0,0 +1,43 @@
+pragma solidity ^0.8.9;
+
+import "../challenge/IChallengeManager.sol";
+import "../rollup/IRollupAdmin.sol";
+
+import "@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+contract EspressoMigration{ 
+    function perform(
```

Comment:
In https://github.com/OffchainLabs/orbit-actions/blob/dc54bae1c40faefb4cdfff82af321cccb9b89a6e/contracts/parent-chain/contract-upgrades/NitroContracts1Point2Point1UpgradeAction.sol#L31

There are some "constants" that are set in the constructor instead of having everything passed to `perform`. I think this makes sense because someone can then deploy then deploy a correctly configured action contract (setting the correct wasm root and OSP implementation) and the the rollup owner only needs to pass what is custom to the rollup.

Reply:
I agree that we should put as much into the constructor as possible. I suppose the question then becomes what constants we can set in the constructor?  To me it seems like we could set the new OSP implememtation address in the constructor, and I'm tempted to say the wasmModuleRoot.

However, there is a case that has me concerned. What if an orbit chain has some pre-existing changes to their STF that they want to merge with ours? If we declare an immutable wasmModuleRoot in the constructor, suddenly we can't use the deployed contract to help upgrade this hypothetical orbit chain as the `wasmModuleRoot` might be different when merging their changes with ours.

Reply:
That rollup would then have to deploy their own upgrade action where they set the root as needed.